biz.aQute.launchpad/src/aQute/launchpad/StandaloneLaunchpadBuilder.java

```diff
@@ -0,0 +1,290 @@
+package aQute.launchpad;
+
+import static aQute.bnd.osgi.Constants.EXPORT_CONTENTS;
+import static aQute.bnd.osgi.Constants.SERVICE_COMPONENT;
+import static aQute.lib.exceptions.Exceptions.unchecked;
+
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.jar.Manifest;
+import java.util.stream.Collectors;
+
+import aQute.bnd.build.Container;
+import aQute.bnd.build.Project;
+import aQute.bnd.build.Run;
+import aQute.bnd.header.Attrs;
+import aQute.bnd.header.Parameters;
+import aQute.bnd.osgi.About;
+import aQute.bnd.osgi.Builder;
+import aQute.bnd.osgi.Descriptors.PackageRef;
+import aQute.bnd.osgi.Jar;
+import aQute.bnd.osgi.Processor;
+import aQute.bnd.service.Strategy;
+import aQute.bnd.service.remoteworkspace.RemoteWorkspace;
+import aQute.bnd.service.specifications.BuilderSpecification;
+import aQute.bnd.service.specifications.RunSpecification;
+import aQute.bnd.version.Version;
+import aQute.bnd.version.VersionRange;
+import aQute.lib.utf8properties.UTF8Properties;
+import aQute.libg.glob.Glob;
+
+public class StandaloneLaunchpadBuilder extends AbstractLaunchpadBuilder<StandaloneLaunchpadBuilder> {
+
+	private final Run run;
+
+	public StandaloneLaunchpadBuilder(File bndrunFile) {
+		this(bndrunFile, unchecked(() -> Run.createRun(null, bndrunFile)));
+	}
+
+	StandaloneLaunchpadBuilder(File bndrunFile, Run run) {
+		super(bndrunFile.getParentFile(), new AbstractRemoteWorkspace(run));
+		this.run = run;
+		RunSpecification analyzedTestSetup = remoteWorkspace.analyzeTestSetup(projectDir.getAbsolutePath());
+		analyzedTestSetup.mergeWith(run.getSpecification());
+		local.mergeWith(analyzedTestSetup);
+	}
+
+	public Run getRun() {
+		return run;
+	}
+
+	static class AbstractRemoteWorkspace implements RemoteWorkspace {
+
+		private final Run run;
+
+		public AbstractRemoteWorkspace(Run run) {
+			this.run = run;
+		}
+
+		@Override
+		public String getBndVersion() {
+			return About.CURRENT.toString();
+		}
+
+		@Override
+		public RunSpecification getRun(String pathToBndOrBndrun) {
+			return unchecked(() -> {
+				File file = new File(pathToBndOrBndrun);
+
+				if (file.isDirectory()) {
+					Project project = run.getWorkspace()
```

Comment:
I went out of my way to not require bnd on the classpath for the test code because it creates to many dependencies. Notice that this contains a lot of code: SLF4j, all the resolve and repo code, etc. The whole idea is that the code runs on the process that launches the test process, not inside the test process. This is a big no-no for me. Launchpad should keep the Workspace and bnd Project code all on the launcher process and never in the launched process.

Reply:
(note that when I say gradle I mean non-workspace `biz.aQute.bnd.builder`)

@pkriens it's all good. This is an experiment. I didn't expect it to be so easy. I just wanted to get my hands dirty to see what could be done.

However, a couple of points:
- there is no stable workspace in maven/gradle to link to
- there are no _projects_ in bnd project terms
- everything in maven/gradle is modeled around Run/Bndrun. The workspace is merely an internal detail of the model rather the main concern.

Reply:
I understand there is no _stable_ workspace, but there **must** be a workspace object. If you look deeper, then you see that the Run _isa_ Project and therefore the Workspace object is absolutely mandatory. A standalone bndrun fakes a workspace as far as I know.

So you should link to this temporary Workspace and let it do the heavy lifting.

Reply:
this would mean making N server backend come up per build!

Reply:
Why remote to something that is instantiated on demand, in process?

Reply:
a) They can share the same backend
b) You cannot do it in the same process because you do not want the whole bnd code on your test class path. 
c) Multiple runs can be faster since you only need to do it once per test process

Reply:
I fail to see who will orchestrate starting this remote process! There is zero bnd code executed until the builder instance is created in the test class at which point starting a remote process is really fruitless!

Furthermore, if what we need is classloader isolation, then let's just do that!

Reply:
Well, on the maven side you 'somewhere' have the bndrun file? That implies you have a workspace?

Reply:
but that workspace is created only when the bndrun file is interpreted which doesn't happen until the test case class is loaded.

Reply:
> Furthermore, if what we need is classloader isolation, then let's just do that!

Would this work, @pkriens? Are there any use cases that you've encountered where (eg) implicitly adding a few extra "excludeExport()" directives to exclude Bnd-related packages wouldn't solve the problem of having the whole bnd code on your test class path?

Reply:
I am pretty sure you're gone get in lots of conflicts. 

I guess the detection code could be instructed with a system property to start a maven workspace for maven in another class loader that has all the bnd code. They could communicate over the remote workspace interface. I guess the analysis code will be  quite different. Maybe we should have a discussion about this.